py-polars/polars/expr/datetime.py

```diff
@@ -2162,6 +2162,192 @@ def month_end(self) -> Expr:
+            ┌─────────────────────┐
+            │ dates               │
+            │ ---                 │
+            │ datetime[μs]        │
+            ╞═════════════════════╡
+            │ 2000-03-31 02:00:00 │
+            │ 2000-03-31 02:00:00 │
+            │ 2000-03-31 02:00:00 │
+            │ 2000-06-30 02:00:00 │
+            │ 2000-06-30 02:00:00 │
+            │ …                   │
+            │ 2000-09-30 02:00:00 │
+            │ 2000-09-30 02:00:00 │
+            │ 2000-12-31 02:00:00 │
+            │ 2000-12-31 02:00:00 │
+            │ 2000-12-31 02:00:00 │
+            └─────────────────────┘
+        """
+        return wrap_expr(self._pyexpr.dt_quarter_end())
+
+    def year_start(self) -> Expr:
+        """
+        Roll backward to the first day of the year.
+
+        Returns
+        -------
+        Expr
+            Expression of data type :class:`Date` or :class:`Datetime`.
+
+        Notes
+        -----
+        If you're coming from pandas, you can think of this as a vectorised version
+        of `pandas.tseries.offsets.YearBegin().rollback(datetime)`.
+
+        Examples
+        --------
+        >>> from datetime import datetime
+        >>> df = pl.DataFrame(
+        ...     {
+        ...         "dates": pl.datetime_range(
+        ...             datetime(2000, 1, 15, 2),
+        ...             datetime(2002, 12, 15, 2),
+        ...             "6mo",
+        ...             eager=True,
+        ...         )
+        ...     }
+        ... )
+        >>> df.select(pl.col("dates").dt.year_start())
+            shape: (6, 1)
+            ┌─────────────────────┐
+            │ dates               │
+            │ ---                 │
+            │ datetime[μs]        │
+            ╞═════════════════════╡
+            │ 2000-01-01 02:00:00 │
+            │ 2000-01-01 02:00:00 │
+            │ 2001-01-01 02:00:00 │
+            │ 2001-01-01 02:00:00 │
+            │ 2002-01-01 02:00:00 │
+            │ 2002-01-01 02:00:00 │
+            └─────────────────────┘
+        """
+        return wrap_expr(self._pyexpr.dt_year_start())
+
+    def year_end(self) -> Expr:
+        """
+        Roll forward to the last day of the year.
+
+        Returns
+        -------
+        Expr
+            Expression of data type :class:`Date` or :class:`Datetime`.
+
+        Notes
+        -----
+        If you're coming from pandas, you can think of this as a vectorised version
+        of `pandas.tseries.offsets.YearEnd().rollforward(datetime)`.
+
+        Examples
+        --------
+        >>> from datetime import datetime
+        >>> df = pl.DataFrame(
+        ...     {
+        ...         "dates": pl.datetime_range(
+        ...             datetime(2000, 1, 15, 2),
+        ...             datetime(2002, 12, 15, 2),
+        ...             "6mo",
+        ...             eager=True,
+        ...         )
+        ...     }
+        ... )
+        >>> df.select(pl.col("dates").dt.year_end())
```

Comment:
can we use `with_columns` for these examples? I realise `month_end` already uses `select` - do you want to change that to `with_columns` first in a precursor PR?

Reply:
I added it to this PR. 
Do you specifically want to split it off?